src/librustc_typeck/astconv.rs

```diff
@@ -944,7 +944,7 @@ impl<'o, 'gcx: 'tcx, 'tcx> AstConv<'gcx, 'tcx>+'o {
 
         debug!("qpath_to_ty: trait_ref={:?}", trait_ref);
 
-        self.normalize_ty(span, tcx.mk_projection(trait_ref, item_segment.name))
+        self.normalize_ty(span, tcx.mk_deprecated_projection(trait_ref, item_segment.name))
```

Comment:
Oh you may have to change the callers of `qpath_to_ty` to pass in the associated type DefId.